feat: add tooling for generating text-only packages

oleksandr-didyk opened the pull request:
## Overview

Resolves https://github.com/dotnet/source-build/issues/2655 

This PR adds tooling for generating text-only packages. The entrypoint for the tooling is the newly added `./generate-text-packs.sh` script, that allows the user to provide a package name, multiple package versions and multiple additional NuGet feeds to restore the package from.

The tooling is a modified version of similar MSBuild logic from `dotnet/installer` ([source](https://github.com/dotnet/installer/blob/6d737887ecc4a3d68f4daede7a2ab021bc67274e/src/SourceBuild/Arcade/tools/SourceBuildArcadeTarball.targets#L342))

The tool was also used to re-generate almost all of the currently present text-only packages

## Other changes

  - renamed `./generate.sh` to `./generate-ref-packs.sh` to better represent its functionality and not collide with the newly added `./generate-text-packs.sh`
  - added missing `*.csproj` files for `Microsoft.CodeAnalysis.Collections` and `Microsoft.NetCore.Platform`
  - replaced existing `LICENSE` files with `dotnet_library_license` (change coming from the NuGet packages themselves)
  - updated `<licenseUrl>` in `*.nuspec` (change coming from the NuGet packages themselves)

## Open questions

  - are the license changes valid or should they be reverted?
  - is the `AllowedTextOnlyExtensions` list valid or should be updated / removed completely? Originally copied from `dotnet/installer` with small additions of the following extensions: `.-`, `.resx`, `.rtf` and `.xlf`
  - existing `*.nuspec` files differ from those available in the NuGet packages by a few lines specified below. Is this something that should be made part of the tooling or could be added manually by the user?
  ```xml
    <files>
      <file src="*/**" target="/" exclude="obj/**/*.*" />
      <file src="*" target="/" />
    </files>
  ```

omajid commented:
> are the license changes valid or should they be reverted?

If the licenses are actually proprietary, that makes it impossible for open source projects to consume this repo and source-build. What would re-licensing look like?

oleksandr-didyk commented on src/textOnlyPackageGenerator/PackageGenerator.proj:
```diff
@@ -0,0 +1,121 @@
+		/>
+	</Target>
+	
+	<!-- Modified version of a similar target in dotnet/installer: -->
```
Sure, wanted to refer to original work, will remove

oleksandr-didyk commented on src/textOnlyPackageGenerator/PackageGenerator.proj:
```diff
@@ -0,0 +1,121 @@
+	
+	<!-- see ./generate-text-packs.sh for `AfterTargets` justification -->
+	<Target Name="CopyTextOnlyPackages" AfterTargets="Restore">
+		<MSBuild 
```
Sure, will fix

oleksandr-didyk commented on src/textOnlyPackageGenerator/PackageGenerator.proj:
```diff
@@ -0,0 +1,121 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Sure, my bad, will fix

oleksandr-didyk commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
```
I mostly kept them separate as to not edit the existing options for ref-packs generation. They are a bit more extensive (providing the option to specify TFM + to read execution options from `.csv`) which were not part of the original issue + I cannot assess their usefulness for text-only packages. On the other hand, there are only two scripts with minimal functionality which I don't think will see much change in terms of available options. With these two factors in mind, I kept the two scripts separate

If need be, I can unify their options and extend / move functionality as we see fit. Don't have a strong opinion on it

MichaelSimons replied:
I realize when we started down this journey it started as bringing in the existing installer logic but the more I look at this, it should be incorporated into the existing tooling rather than two distinct paths.  Beyond sharing the options the logic to restore the packages should be shared across both types as well.

oleksandr-didyk replied:
Sure, so to outline:

  - reference package and text-only package generation options should be unified (user can either manually specify package name and version together with optional inputs like TFM or feeds or they can provide a list of options to execute through a `.csv` or alike)
  - common code should be shared between the two projects (a good example is the restore procedure)
  - unlike existing reference package generation logic, the target one should behave the same way as the text-only one -> a thin wrapper delegating work to the project (point you made in your original comment)

oleksandr-didyk commented:
Update - incorporating PR feedback: 

- removed reference to `dotnet/installer` in target comment
- fixed spacing and formatting in `PackageGenerator.proj`

premun commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+			usage
+			exit 0
+			;;
+		-name|-n)
```
I can see the other script using `--` for long argument names (I think this is the standard expectation?)

oleksandr-didyk replied:
This script does as well, it just trims the double dash on line 45
```bash
"${1/#--/-}"
```

premun commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+fi
+
+# -p:RestoreAdditionalProjectSources -> specifies additional Nuget package sources, including feeds: https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets#restore-properties
+# -p:RestoreUsingNuGetTargets -> switch that controls who will restore the project; if set to `false`, Arcade's Build.proj will execute the Restore target of the project itself.
```
We could, theoretically, also set the `RestoreUsingNuGetTargets` property inside of the `.proj` file and this way we embed this silent contract of how the proj file needs to be run inside of it. Would that work?

Something like
```xml
<RestoreUsingNuGetTargets Condition=" '$(RestoreUsingNuGetTargets)' != '' ">false</RestoreUsingNuGetTargets>
```

oleksandr-didyk replied:
Someone can brake it by accident by explicitly setting `/p:RestoreUsingNuGetTargets=true`. Furthermore, this option is only needed if you utilize Arcade's `build/sh`. If someone were to run the project via `MSBuild` directly, it would work without the property, making the assignment inside the `.proj` file redundant

Since we only have 1 entry-point and the property is specific to this entry-point and not to the project itself, I would keep it here

premun replied:
Sure, if we always keep the `.proj` with the `.sh` it makes sense so hopefully they will.

crummel commented:
> > are the license changes valid or should they be reverted?
> 
> If the licenses are actually proprietary, that makes it impossible for open source projects to consume this repo and source-build. What would re-licensing look like?

I believe Dan S went and got permission to put these under the licenses as currently checked-in so we shouldn't revert.  @leecow can probably confirm.

MichaelSimons commented:
> I believe Dan S went and got permission to put these under the licenses as currently checked-in so we shouldn't revert. @leecow can probably confirm.

I confirmed this with Dan S.

tkapin commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+# resolve $SOURCE until the file is no longer a symlink
+while [[ -h $source ]]; do
+  scriptroot="$( cd -P "$( dirname "$source" )" && pwd )"
+  source="$(readlink "$source")"
```
Can you check if readlink -f (the canonical resolution) would do the trick?

oleksandr-didyk replied:
Seems to be working fine with `readlink -f`, I'll use it instead, thanks for the pointer

tkapin commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+source="${BASH_SOURCE[0]}"
+
+# resolve $SOURCE until the file is no longer a symlink
+while [[ -h $source ]]; do
```
Always use doublequotes in tests as otherwise the condition would become syntactically invalid when the variable would be empty.

premun replied:
I recommend VS Code + Shellcheck:
https://marketplace.visualstudio.com/items?itemName=timonwong.shellcheck

tkapin commented on generate-text-packs.sh:
```diff
@@ -0,0 +1,85 @@
+fi
+
+while [[ $# > 0 ]]; do
+	opt="$(echo "${1/#--/-}" | tr "[:upper:]" "[:lower:]")"
```
Linux command options are commonly case sensitive, we shouldn't be manipulating case of the options.

oleksandr-didyk commented:
Update - incorporating PR feedback:

  - removed changes to package licenses
  - unified generator scripts
  - extracted ref-pack restore logic and re-used it for text-only packages
  - minor improvements to the `./generate.sh` script

premun commented on generate.sh:
```diff
@@ -27,10 +16,13 @@ NC='\033[0m'
-    echo "  to build a reference package.  The generated artifacts are added to the /src/ directory of this repo"
-    echo "  unless a different destination is specified.  If a csv file of packages is specified, multiple"
-    echo "  packages and their dependencies are generated."
+    echo "  Generates a reference package or a text-only package from the specified packages and versions. The"
+    echo "  type of the generated package is controlled via the --textOnly option."
+    echo ""
+    echo "  Reference package generation will restores reference package(s) and dependencies and generates cs files"
+    echo "  and accompanying projects into the specified destination ('./src/referencePackages/' by default)."
+    echo "  Text-only package generation will restore the specified package and copy the source-build-usable content"
+    echo "  into the provided directory ('./src/textOnlyPackages/' by default)."
```
```suggestion
    echo "  Generates a reference package or a text-only package from specified packages and versions. The"
    echo "  type of the generated package is controlled via the --textOnly option."
    echo ""
    echo "  Reference package generation will restore reference package(s) and dependencies and generate .cs files"
    echo "  with accompanying projects into the specified destination ('./src/referencePackages/' by default)."
    echo "  Text-only package generation will restore the specified package and copy the source-build-usable content"
    echo "  into the provided directory ('./src/textOnlyPackages/' by default)."
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on generate.sh:
```diff
@@ -42,21 +34,24 @@ usage() {
     echo "                                       option above, one per line.  If specified, the --pkg option is ignored."
     echo "  --dest <pathToDestRepo>            A path to the root of the repo to copy source into."
+    echo "  --textOnly                         A switch to generate text-only packages."
+    echo "  --feeds <nugetFeeds>               A semicolon-separated list of additional Nuget feeds to use during restore."
```
```suggestion
    echo "  --feeds <nugetFeeds>               A semicolon-separated list of additional NuGet feeds to use during restore."
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on src/common/RestoreProjects.proj:
```diff
@@ -0,0 +1,18 @@
+  <Target Name="ErrorEmptyRestoreProjects"
+    BeforeTargets="RestoreAllProjects"
+    Condition=" '@(RestoreProjects)' == '' ">
+    <Error Text="There are no restore projects to process." />
```
```suggestion
    <Error Text="There are no projects to restore." />
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on src/common/RestoreProjects.proj:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project>
+  <ItemGroup>
+    <RestoreProjects Include="$(RepoRoot)artifacts/targetRestoreProjects/*.csproj"/>
```
```suggestion
    <RestoreProjects Include="$(RepoRoot)artifacts/targetRestoreProjects/*.csproj" />
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on src/textOnlyPackageGenerator/PackageGenerator.proj:
```diff
@@ -0,0 +1,118 @@
+    </ItemGroup>
+  </Target>
+    
+  <!-- see ./generate-text-packs.sh for `AfterTargets` justification -->
```
```suggestion
  <!-- see ./generate.sh for `AfterTargets` justification -->
```

premun commented on generate.sh:
```diff
@@ -130,21 +131,41 @@ do
-./CopyProjects.sh "$pathToRepoSrc"
-popd
+    pushd "$pathToGeneratedSrc"
+    chmod 755 ./CopyProjects.sh
```
```suggestion
    chmod +x ./CopyProjects.sh
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on generate.sh:
```diff
@@ -42,21 +34,24 @@ usage() {
     echo "  --pkgCsv <pathToCSV>               A path to a csv file of packages to generate. Format is the same as the --pkg"
     echo "                                       option above, one per line.  If specified, the --pkg option is ignored."
     echo "  --dest <pathToDestRepo>            A path to the root of the repo to copy source into."
+    echo "  --textOnly                         A switch to generate text-only packages."
```
```suggestion
    echo "  --textOnly                         Generate text-only packages (skip building)."
```

oleksandr-didyk replied:
Fixed, thanks

premun commented on src/textOnlyPackageGenerator/PackageGenerator.proj:
```diff
@@ -0,0 +1,118 @@
+    <ItemGroup>
+      <TextOnlyPackageContent
+        Include="$(RestoredPackagePath)/**/*"
+        Exclude="
+          $(RestoredPackagePath)/**/.nupkg.metadata;
+          $(RestoredPackagePath)/**/.signature.p7s;
+          $(RestoredPackagePath)/**/*.nupkg;
+          $(RestoredPackagePath)/**/*.nupkg.sha512;"/>
```
Why do we exclude some here and some below?

oleksandr-didyk replied:
Here we explicitly remove metadata assets we know we don't need. Below we verify that there are no assets in the package that we don't know of. If any are found - we raise an exception. We don't remove any file there

MichaelSimons commented on generate.sh:
```diff
@@ -130,21 +131,41 @@ do
-pathToRepoSrc="$pathToDestRepo/src/referencePackages/src"
+    # Copy source to destination
+    pathToGeneratedSrc="$scriptroot/artifacts/generatedSrc/"
+    if [ -z "$pathToDestRepo" ]; then
```
Perhaps the type option I mentioned earlier would make it easier to share this logic to set the pathToDestRepo which is similar across ref and text-only packages.

oleksandr-didyk replied:
Yeah, that's true. Will change it together with the `type` option

MichaelSimons reviewed (commented):
Changes look nice.  I just have a couple comments.

oleksandr-didyk commented on generate.sh:
```diff
@@ -78,9 +72,15 @@ while :; do
             pathToDestRepo="$(cd -P "$2" && pwd)"
             shift
             ;;
+        --textOnly)
```
That's a great point, thanks for bringing it up. Will change

oleksandr-didyk commented on generate.sh:
```diff
@@ -130,21 +131,41 @@ do
-chmod 755 ./CopyProjects.sh
-./CopyProjects.sh "$pathToRepoSrc"
-popd
+    pushd "$pathToGeneratedSrc"
```
Agree, thanks for pointing it out. Will change

MichaelSimons commented:
@oleksandr-didyk, When making changes in response to code review feedback, I find it beneficial to preserve the individual commits.  It helps the code reviewers see what was changed versus having to re-review the entire set of changes.  The commits can be squashed via GitHub when merging the PR.

MichaelSimons commented on generate.sh:
```diff
@@ -42,21 +33,25 @@ usage() {
     echo "  --pkgCsv <pathToCSV>               A path to a csv file of packages to generate. Format is the same as the --pkg"
     echo "                                       option above, one per line.  If specified, the --pkg option is ignored."
     echo "  --dest <pathToDestRepo>            A path to the root of the repo to copy source into."
+    echo "  --type <packageType>               Type of the package to generate. Accepted values: text | ref."
```
The ending punctuation is confusing IMO especially when it is included the the next line that indicates the default.  It can leave the user with the impression the value they must specify is actually `ref.`
```suggestion
    echo "  --type <packageType>               Type of the package to generate. Accepted values: ref (default) | text"
```

oleksandr-didyk replied:
Sure, thanks for pointing it out, will change
